### PR TITLE
ENH: Add nib-roi command to crop (maybe flip) axes

### PR DESCRIPTION
**Pull request by effigies**
Quick command to easily crop images using slice notation. Inspired by `fslroi`.

Features:

1) Python slicing syntax.
2) Scale factors and original values preserved exactly.
3) Flips on spatial axes are permitted with a `-1` step (but downsampling with other steps is not...)

Safety checks ~~that are not there:~~

1) Zero-length slices are ~~not~~ checked for, e.g. `-i 1:1` or `-j 5:6:-1`.
2) Using the same input and output file on an uncompressed image ~~will cause a BusError as the image being read will be truncated.~~ will avoid mmaping and is tested for.

With a basic usable tool, I'm not inclined to push further today, so I figured I'd give people a chance to comment.

### CODE REVIEW

**Comment by effigies** [comment]
This could use a review. @chrispycheng any interest?

**Comment by chrispycheng** [comment]
@effigies Sure, let me get back to you! Plate's a bit stacked ATM

**Comment by effigies** [comment]
No worries.

File: nibabel/cmdline/tests/test_roi.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+@pytest.mark.parametrize("args, errmsg", (
+    (("-i", "1:1"), "Cannot take zero-length slice"),
+    (("-j", "1::2"), "Downsampling is not supported"),
+    (("-t", "5::-1"), "Step entry not permitted"),
+))
+def test_nib_roi_bad_slices(capsys, args, errmsg):
+    in_file = os.path.join(data_path, 'functional.nii')
+
+    retval = main([in_file, '/dev/null', *args])
```
Comment (chrispycheng) [inline nibabel/cmdline/tests/test_roi.py:111]: Will '/dev/null' work for Windows as well?

**Review by chrispycheng — COMMENTED** [review]
Ran the function and looks awesome! Skimmed the code and raised just one point of concern for your review. Also, figured I wouldn't rehash codecov, esp. since areas lacking coverage seem pretty insignificant.

File: nibabel/cmdline/tests/test_roi.py (hunk 1 above)
Comment (effigies) [reply nibabel/cmdline/tests/test_roi.py:111]: Probably, but this should be `os.devnull`.

```suggestion
    retval = main([in_file, os.devnull, *args])
```

**Comment by effigies** [comment]
Thanks for the review. I went ahead and added tests that should cover the remaining lines.

**Comment by effigies** [comment]
The quest for coverage ended up finding an edge case. Full coverage and tests passing.

@chrispycheng Do you have time for a final review?

**Comment by chrispycheng** [comment]
I’m smothered with work right now.. maybe in 2 weeks? is that OK?

Sent from my iPhone

On Oct 9, 2020, at 21:17, Chris Markiewicz <notifications@github.com> wrote:

﻿

The quest for coverage ended up finding an edge case. Full coverage and tests passing.

@chrispycheng<https://github.com/chrispycheng> Do you have time for a final review?

**Comment by effigies** [comment]
No worries. Thanks for the heads up.

File: nibabel/cmdline/tests/test_roi.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+from nibabel.cmdline.roi import lossless_slice, parse_slice, main
+from nibabel.testing import data_path
+
+import unittest
+from unittest import mock
+import pytest
+
+
+def test_parse_slice():
+    assert parse_slice(None) == slice(None)
```
Comment (mgxd) [inline nibabel/cmdline/tests/test_roi.py:13]: where is `slice` defined?

File: nibabel/cmdline/roi.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    kwargs = {}
+    if os.path.realpath(opts.in_file) == os.path.realpath(opts.out_file):
+        kwargs['mmap'] = False
+    img = nb.load(opts.in_file, **kwargs)
+
+    slicers = (islice, jslice, kslice, tslice)[:img.ndim]
+    expected_shape = nb.fileslice.predict_shape(slicers, img.shape)
+    if any(dim == 0 for dim in expected_shape):
+        print(f"Cannot take zero-length slices. Predicted shape {expected_shape}.")
+        return 1
```
Comment (mgxd) [inline nibabel/cmdline/roi.py:80]: is this passed along to the command line?
```suggestion
        sys.exit(1)
```

File: nibabel/cmdline/roi.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    try:
+        sliced_img = lossless_slice(img, slicers)
+    except Exception:
+        print("Could not slice image. Full traceback follows.")
+        raise
+    nb.save(sliced_img, opts.out_file)
+    return 0
```
Comment (mgxd) [inline nibabel/cmdline/roi.py:88]: is this needed?
```suggestion
```

File: nibabel/cmdline/roi.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    parser.add_argument("-i", metavar="I1:I2[:-1]",
+                        help="Start/stop [flip] along first axis (0-indexed)")
+    parser.add_argument("-j", metavar="J1:J2[:-1]",
+                        help="Start/stop [flip] along second axis (0-indexed)")
+    parser.add_argument("-k", metavar="K1:K2[:-1]",
+                        help="Start/stop [flip] along third axis (0-indexed)")
+    parser.add_argument("-t", metavar="T1:T2", help="Start/stop along fourth axis (0-indexed)")
+    parser.add_argument("in_file", help="Image file to crop")
+    parser.add_argument("out_file", help="Output file name")
+
```
Comment (mgxd) [inline nibabel/cmdline/roi.py:59]: should this show the help if no arguments are provided?
```suggestion
    if not args:
        parser.print_help()
        sys.exit(1)
```

**Review by mgxd — COMMENTED** [review]
Looks good - few comments

File: nibabel/cmdline/tests/test_roi.py (hunk 2 above)
Comment (effigies) [reply nibabel/cmdline/tests/test_roi.py:13]: Python builtin.

File: nibabel/cmdline/roi.py (hunk 4 above)
Comment (effigies) [reply nibabel/cmdline/roi.py:88]: Not really, makes the output of calling `main()` identical to the exit code, though. If I change this, I think I'll need to update a test, but I can.

File: nibabel/cmdline/roi.py (hunk 3 above)
Comment (effigies) [reply nibabel/cmdline/roi.py:80]: Yes.

File: nibabel/cmdline/roi.py (hunk 4 above)
Comment (mgxd) [reply nibabel/cmdline/roi.py:88]: nbd 🤷‍♂️

File: nibabel/cmdline/roi.py (hunk 5 above)
Comment (effigies) [reply nibabel/cmdline/roi.py:59]: You do at least get usage:

```
$ nib-roi
usage: nib-roi [-h] [--version] [-i I1:I2[:-1]] [-j J1:J2[:-1]]
               [-k K1:K2[:-1]] [-t T1:T2]
               in_file out_file
nib-roi: error: the following arguments are required: in_file, out_file
$ nib-roi --help
usage: nib-roi [-h] [--version] [-i I1:I2[:-1]] [-j J1:J2[:-1]]
               [-k K1:K2[:-1]] [-t T1:T2]
               in_file out_file

Crop images to a region of interest

positional arguments:
  in_file        Image file to crop
  out_file       Output file name

optional arguments:
  -h, --help     show this help message and exit
  --version      show program's version number and exit
  -i I1:I2[:-1]  Start/stop [flip] along first axis (0-indexed)
  -j J1:J2[:-1]  Start/stop [flip] along second axis (0-indexed)
  -k K1:K2[:-1]  Start/stop [flip] along third axis (0-indexed)
  -t T1:T2       Start/stop along fourth axis (0-indexed)

If a start or stop value is omitted, the start or end of the axis is assumed.
```

We don't have a terribly consistent style, to go by:

```
$ nib-ls; echo $?

0
$ nib-conform; echo $?
usage: nib-conform [-h] [--out-shape OUT_SHAPE OUT_SHAPE OUT_SHAPE]
                   [--voxel-size VOXEL_SIZE VOXEL_SIZE VOXEL_SIZE]
                   [--orientation ORIENTATION] [-f] [-V]
                   infile outfile
nib-conform: error: the following arguments are required: infile, outfile
2
$ nib-diff; echo $?
Traceback (most recent call last):
  File "/home/cjmarkie/anaconda3/envs/nibabel_dev/bin/nib-diff", line 11, in <module>
    load_entry_point('nibabel', 'console_scripts', 'nib-diff')()
  File "/home/cjmarkie/Projects/nipy/nibabel/nibabel/cmdline/diff.py", line 362, in main
    dtype=opts.dtype
  File "/home/cjmarkie/Projects/nipy/nibabel/nibabel/cmdline/diff.py", line 315, in diff
    assert len(files) >= 2, "Please enter at least two files"
AssertionError: Please enter at least two files
1
$ nib-tck2trk; echo $?
usage: nib-tck2trk [-h] [-f] anatomy tractogram [tractogram ...]
nib-tck2trk: error: the following arguments are required: anatomy, tractogram
2
```

But on the whole I think I'm okay with argparse defaults.

**Comment by effigies** [comment]
Since none of the comments seemed blocking, I'm going to go ahead and merge. Thanks for the reviews, @chrispycheng, @mgxd!
